File: embedded_C/unittest/platform_test/workers_test.cpp
#include <atomic>
#include <chrono>
#include <memory>
#include <optional>
#include <stdexcept>
#include <string>
#include <thread>

#include "1_Platform/message_bus/message_bus.hpp"
#include "3_Devices/lidars/lidar.hpp"
#include "4_Applications/processing/pointcloud/pointcloud_processing.hpp"
#include "models/topics.hpp"
#include "unittest/test.hpp"

namespace {

class MockReader final : public vista::devices::ILidarReader {
public:
    explicit MockReader(vista::platform::StopToken stop) : stop_(std::move(stop)) {}

    vista::devices::RawPacket read_raw_packet() override {
        std::this_thread::sleep_for(std::chrono::milliseconds(2));
        stop_.request_stop();
        return vista::devices::RawPacket({1, 2, 3}, 100);
    }

private:
    vista::platform::StopToken stop_;
};

class MockDecoder final : public vista::devices::ILidarDecoder {
public:
    vista::models::PointCloudFrame decode_packet(
        const vista::devices::RawPacket& packet) override {
        VISTA_CHECK(packet.bytes() == std::vector<std::uint8_t>({1, 2, 3}));
        return {
            packet.timestamp_ns().value_or(0),
            {vista::models::PointXYZIRT{1.0F, 0.0F, 0.5F, 1, 0, 0, 100}},
        };
    }
};

}  // namespace

VISTA_TEST(workers_register_their_own_topics_and_process_one_packet) {
    vista::platform::StopToken stop;
    vista::platform::MessageBus bus;
    bus.configure_topic(vista::models::topics::lidar_raw, 2);
    bus.configure_topic(vista::models::topics::pointcloud_decoded, 2);
    bus.configure_topic(vista::models::topics::pointcloud_processed, 2);

    std::optional<vista::devices::LidarWorkerReport> read_report;
    std::optional<vista::devices::LidarWorkerReport> decode_report;
    std::optional<vista::application::PreprocessingReport> preprocessing_report;
    std::string read_error;
    std::string decode_error;
    std::string preprocessing_error;
    auto decoder_slot =
        std::make_shared<vista::devices::LidarDecoderSlot>();

    // Consumers are started first, but each worker performs its own subscribe
    // and publisher setup rather than receiving endpoints from this test.
    auto preprocessing = vista::application::spawn_preprocessing_worker(
        bus,
        vista::platform::ThreadConfig("test-preprocessing", 3),
        stop,
        vista::application::PreprocessingConfig{
            0.0F, 10.0F, std::nullopt, std::nullopt, std::nullopt},
        [&](auto report, auto error) {
            preprocessing_report = std::move(report);
            preprocessing_error = std::move(error);
        });
    auto decoder = vista::devices::spawn_lidar_decode_worker(
        bus,
        vista::platform::ThreadConfig("test-lidar-decode", 2),
        stop,
        decoder_slot,
        [&](auto report, auto error) {
            decode_report = std::move(report);
            decode_error = std::move(error);
        });
    auto reader = vista::devices::spawn_lidar_read_worker(
        bus,
        vista::platform::ThreadConfig("test-lidar-read", 1),
        stop,
        [stop]() {
            return vista::devices::Lidar::from_parts(
                "Mock LiDAR",
                "mock-lidar",
                std::make_unique<MockReader>(stop),
                std::make_unique<MockDecoder>());
        },
        std::chrono::milliseconds(1),
        decoder_slot,
        [&](auto report, auto error) {
            read_report = std::move(report);
            read_error = std::move(error);
        });

    reader.join();
    decoder.join();
    preprocessing.join();

    VISTA_CHECK(read_error.empty());
    VISTA_CHECK(decode_error.empty());
    VISTA_CHECK(preprocessing_error.empty());
    VISTA_CHECK(read_report && read_report->message_count == 1);
    VISTA_CHECK(decode_report && decode_report->message_count == 1);
    VISTA_CHECK(
        preprocessing_report && preprocessing_report->message_count == 1);
    VISTA_CHECK(preprocessing_report->point_count == 1);
}

VISTA_TEST(lidar_read_worker_retries_after_connection_failure) {
    vista::platform::StopToken stop;
    vista::platform::MessageBus bus;
    bus.configure_topic(vista::models::topics::lidar_raw, 2);

    std::atomic<int> connection_attempts{0};
    std::optional<vista::devices::LidarWorkerReport> read_report;
    std::string read_error;
    auto decoder_slot =
        std::make_shared<vista::devices::LidarDecoderSlot>();

    auto reader = vista::devices::spawn_lidar_read_worker(
        bus,
        vista::platform::ThreadConfig("test-lidar-reconnect", 1),
        stop,
        [&]() {
            if (++connection_attempts == 1) {
                throw std::runtime_error("mock LiDAR is disconnected");
            }
            return vista::devices::Lidar::from_parts(
                "Mock LiDAR",
                "mock-lidar",
                std::make_unique<MockReader>(stop),
                std::make_unique<MockDecoder>());
        },
        std::chrono::milliseconds(1),
        decoder_slot,
        [&](auto report, auto error) {
            read_report = std::move(report);
            read_error = std::move(error);
        });

    reader.join();

    VISTA_CHECK(connection_attempts == 2);
    VISTA_CHECK(read_error.empty());
    VISTA_CHECK(read_report && read_report->message_count == 1);
}

File: embedded/unittest/platform_test/workers_test.rs
//! Integration tests for independent workers and their self-registered topics.

use std::{
    io,
    sync::{
        atomic::{AtomicUsize, Ordering},
        mpsc, Arc,
    },
    thread,
    time::Duration,
};

use crate::{
    application::pointcloud_processing::{
        spawn_preprocessing_worker, PreprocessingConfig, PreprocessingReport,
    },
    devices::lidar::{
        spawn_lidar_decode_worker, spawn_lidar_read_worker, Lidar, LidarConnector, LidarDecoder,
        LidarDecoderSlot, LidarReader, LidarWorkerReport, RawPacket,
    },
    models::{
        pointcloud::{PointCloudFrame, PointXYZIRT},
        topics,
    },
    platform::{
        message_bus::MessageBus,
        threading::{StopToken, ThreadConfig},
    },
};

struct MockReader {
    stop: StopToken,
}

impl LidarReader for MockReader {
    fn read_raw_packet(&mut self) -> io::Result<RawPacket> {
        thread::sleep(Duration::from_millis(2));
        self.stop.request_stop();
        Ok(RawPacket::new(vec![1, 2, 3]).with_timestamp_ns(100))
    }
}

struct MockDecoder;

impl LidarDecoder for MockDecoder {
    fn decode_packet(&mut self, packet: &RawPacket) -> io::Result<PointCloudFrame> {
        assert_eq!(packet.as_bytes(), &[1, 2, 3]);
        Ok(PointCloudFrame::new(
            packet.timestamp_ns().unwrap_or_default(),
            vec![PointXYZIRT {
                x: 1.0,
                y: 0.0,
                z: 0.5,
                intensity: 1,
                ring: 0,
                return_id: 0,
                timestamp_ns: 100,
            }],
        ))
    }
}

/// Consumers may start first because endpoint registration belongs to workers.
#[test]
fn workers_register_topics_and_process_one_packet() {
    let stop = StopToken::new();
    let bus = Arc::new(MessageBus::new(16).unwrap());
    bus.configure_topic(topics::LIDAR_RAW, 2).unwrap();
    bus.configure_topic(topics::POINTCLOUD_DECODED, 2).unwrap();
    bus.configure_topic(topics::POINTCLOUD_PROCESSED, 2)
        .unwrap();
    let decoder_slot = Arc::new(LidarDecoderSlot::new());
    let (preprocess_tx, preprocess_rx) = mpsc::channel();
    let (decode_tx, decode_rx) = mpsc::channel();
    let (read_tx, read_rx) = mpsc::channel();

    let preprocessing = spawn_preprocessing_worker(
        Arc::clone(&bus),
        ThreadConfig::new("test-preprocessing", 3).unwrap(),
        stop.clone(),
        PreprocessingConfig {
            min_distance_m: 0.0,
            max_distance_m: 10.0,
            region_of_interest: None,
            voxel_size_m: None,
            ground_removal: None,
        },
        move |result: Result<PreprocessingReport, String>| preprocess_tx.send(result).unwrap(),
    )
    .unwrap();
    let decoder = spawn_lidar_decode_worker(
        Arc::clone(&bus),
        ThreadConfig::new("test-lidar-decode", 2).unwrap(),
        stop.clone(),
        Arc::clone(&decoder_slot),
        move |result: Result<LidarWorkerReport, String>| decode_tx.send(result).unwrap(),
    )
    .unwrap();
    let connector_stop = stop.clone();
    let connector: LidarConnector = Arc::new(move || {
        Ok(Lidar::from_parts(
            "Mock LiDAR",
            "mock-lidar",
            Box::new(MockReader {
                stop: connector_stop.clone(),
            }),
            Box::new(MockDecoder),
        ))
    });
    let reader = spawn_lidar_read_worker(
        Arc::clone(&bus),
        ThreadConfig::new("test-lidar-read", 1).unwrap(),
        stop,
        connector,
        Duration::from_millis(1),
        decoder_slot,
        move |result: Result<LidarWorkerReport, String>| read_tx.send(result).unwrap(),
    )
    .unwrap();

    reader.join().unwrap();
    decoder.join().unwrap();
    preprocessing.join().unwrap();
    let read = read_rx.recv().unwrap().unwrap();
    let decode = decode_rx.recv().unwrap().unwrap();
    let preprocessing = preprocess_rx.recv().unwrap().unwrap();
    assert_eq!(read.message_count, 1);
    assert_eq!(decode.message_count, 1);
    assert_eq!(preprocessing.message_count, 1);
    assert_eq!(preprocessing.point_count, 1);
}

/// A failed connection is reported and retried instead of stopping the process.
#[test]
fn lidar_read_worker_retries_after_connection_failure() {
    let stop = StopToken::new();
    let bus = Arc::new(MessageBus::new(16).unwrap());
    bus.configure_topic(topics::LIDAR_RAW, 2).unwrap();
    let decoder_slot = Arc::new(LidarDecoderSlot::new());
    let attempts = Arc::new(AtomicUsize::new(0));
    let connector_attempts = Arc::clone(&attempts);
    let connector_stop = stop.clone();
    let connector: LidarConnector = Arc::new(move || {
        if connector_attempts.fetch_add(1, Ordering::SeqCst) == 0 {
            return Err(io::Error::new(
                io::ErrorKind::NotConnected,
                "mock LiDAR is disconnected",
            ));
        }
        Ok(Lidar::from_parts(
            "Mock LiDAR",
            "mock-lidar",
            Box::new(MockReader {
                stop: connector_stop.clone(),
            }),
            Box::new(MockDecoder),
        ))
    });
    let (result_tx, result_rx) = mpsc::channel();
    let reader = spawn_lidar_read_worker(
        Arc::clone(&bus),
        ThreadConfig::new("test-lidar-reconnect", 1).unwrap(),
        stop,
        connector,
        Duration::from_millis(1),
        decoder_slot,
        move |result: Result<LidarWorkerReport, String>| result_tx.send(result).unwrap(),
    )
    .unwrap();

    reader.join().unwrap();
    let report = result_rx.recv().unwrap().unwrap();
    assert_eq!(attempts.load(Ordering::SeqCst), 2);
    assert_eq!(report.message_count, 1);
}
